File: 2/RiverSizes.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Queue;

class RiverSizes {

	public static List<Integer> getRiverSizes(int matrix[][]) {
		List<Integer> riverSizes = new ArrayList<Integer>();	
		Map<Pair, Boolean> visited = new HashMap<Pair, Boolean>();	
		
		int r = matrix.length;
		int c = matrix[0].length;

		for(int i=0; i<r; i++) {
			for(int j=0; j<c; j++) {
				Pair pair = new Pair(i,j);
				
				if(visited.get(pair) == null) {
					if(matrix[i][j] == 1) {
						int riverSize = 0;
						Queue<Pair> q = new LinkedList<Pair>();
						q.add(pair);	

						while(!q.isEmpty()) {
							Pair cur = q.remove();

							if((cur.x >= 0 && cur.x < r) && (cur.y >= 0 && cur.y < c)) {
								if(visited.get(cur) == null && matrix[cur.x][cur.y] == 1) {
									riverSize++;
									q.add(new Pair(cur.x-1, cur.y));
									q.add(new Pair(cur.x+1, cur.y));
									q.add(new Pair(cur.x, cur.y-1));
									q.add(new Pair(cur.x, cur.y+1));
								}
								visited.put(cur, true);
							}
						}

						if(riverSize != 0) {
							riverSizes.add(riverSize);	
						}
					} else {
						visited.put(pair, true);
					}

				} 
			}
		}

		return riverSizes; 
	}

	public static void main(String args[]) {
		int matrix[][] = { {1,0,0,1,0}, {1,0,1,0,0}, {0,0,1,0,1}, {1,0,1,0,1}, {1,0,1,1,0} }; 
		

		List<Integer> riverSizes = getRiverSizes(matrix);
		
		System.out.println(riverSizes.toString());
	}

	// Helper class
	static class Pair {
		int x,y;

		Pair(int x, int y) {
			this.x = x;
			this.y = y;
		}
		
		@Override
		public int hashCode() {
			return Objects.hash(this.x, this.y);
		}

		@Override
		public boolean equals(Object o) {
			if(o == this) {
				return true;
			}

			if(!(o instanceof Pair)) {
				return false;
			}
			
			Pair pair = (Pair)o;
			
			return pair.x == this.x && pair.y == this.y; 
		}
	}
}

File: 2/js/src/RiverSizes.ts
function getRiverSizes(matrix: number[][]): number[] {
	let visited = matrix.map(row => row.map(() => false));
	let sizes: number[] = [];

	for(let i=0; i<matrix.length; i++) {
		for(let j=0; j<matrix[0].length; j++) {
			if(!visited[i][j]) explore(i, j, matrix, visited, sizes);
		}
	}

	return sizes;
}

function explore(i: number, j: number, matrix: number[][], visited: boolean[][], sizes: number[]) {
	if(matrix[i][j] == 0) {
		visited[i][j] = true;
		return;
	}

	let curSize = 0;

	let queue: number[][] = [[i,j]];

	while(queue.length != 0) {
		const cur = queue.shift()!;
		const [x,y] = cur;

		visited[x][y] = true;

		if(matrix[x][y] == 0) continue;

		curSize++;

		const neighbours = getUnvisitedNeighbours(cur[0], cur[1], matrix, visited);
		for(const neighbour of neighbours) {
			queue.push(neighbour);
		}
	}

	if(curSize > 0) {
		sizes.push(curSize);
	}
}

function getUnvisitedNeighbours(i: number, j: number, matrix: number[][], visited: boolean[][]): number[][] {
	const xDir = [0,0,1,-1];
	const yDir = [1,-1,0,0];

	let neighbours = [];

	for(let k=0; k<4; k++) {
		const x = i + xDir[k];
		const y = j + yDir[k];

		if( (x >= 0 && x < matrix.length) && (y >= 0 && y < matrix[0].length) ) {
			if(!visited[x][y]) {
				neighbours.push([x,y]);
			}
		}
	}

	return neighbours;
}

(function main() {
	const matrix = [ [1,0,0,1,0], [1,0,1,0,0], [0,0,1,0,1], [1,0,1,1,0] ];
	console.log(getRiverSizes(matrix));
})();
